src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -77,6 +78,25 @@ private static Exception GetInvalidArgumentException(string buffer, int offset,
             return ThrowHelper.GetArgumentException(ExceptionResource.Argument_InvalidOffsetLength);
         }
 
+        /// <summary>
+        /// Initializes an instance of the <see cref="StringSegment"/> struct.
+        /// </summary>
+        /// <param name="span">The original <see cref="Span{T}"/> used as buffer.</param>
+        public unsafe StringSegment(Span<char> span)
+        {
+            Buffer = new string((char*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(span)));
```

Comment:
The use of `AsPointer` risks a GC hole as the spans data may not be pinned; you'd have to fix it instead to get the pointer. Also should use an override that takes length (or will have issues with null termination)

```csharp
fixed (char* src = &MemoryMarshal.GetReference(span))
{
    Buffer = new string(src, span.Length);
}
```

Reply:
Thanks for the hint!